src/main/kotlin/org/rust/lang/core/macros/MacroExpansionManager.kt

```diff
@@ -218,32 +230,113 @@ class MacroExpansionManagerImpl(
 }
 
 private fun updateDirs(projectDirName: String?): Dirs {
-    return updateDirs0(projectDirName ?: RandomStringUtils.randomAlphabetic(8))
+    return updateDirs0(projectDirName ?: randomLowercaseAlphabetic(8))
 }
 
 private fun updateDirs0(projectDirName: String): Dirs {
-    val baseProjectDir = getBaseMacroExpansionDir().resolve(projectDirName)
-        .also { it.createDirectories() }
-    val expansionsDir = baseProjectDir.resolve("expansions")
+    val baseProjectDir = getBaseMacroDir()
         .also { it.createDirectories() }
     return Dirs(
-        baseProjectDir,
-        baseProjectDir.resolve("data.dat"),
-        LocalFileSystem.getInstance().refreshAndFindFileByIoFile(expansionsDir.toFile())!!
+        baseProjectDir.resolve("$projectDirName.dat"),
+        projectDirName
     )
 }
 
 private data class Dirs(
-    val baseProjectDir: Path,
     val dataFile: Path,
-    val expansionsDirVi: VirtualFile
-)
+    val projectDirName: String
+) {
+    // Path in the MacroExpansionVFS
+    val expansionDirPath get() = "/$MACRO_EXPANSION_VFS_ROOT/$projectDirName"
+}
 
 private class MacroExpansionServiceImplInner(
     private val project: Project,
     val dirs: Dirs,
-    private val storage: ExpandedMacroStorage
+    private val storage: ExpandedMacroStorage,
+    var expansionsDirVi: VirtualFile?
 ) {
+    companion object {
+        fun build(project: Project, dirs: Dirs): () -> MacroExpansionServiceImplInner {
```

Comment:
I still don't like it
We have a lot of indirection in macro expansion manager initialization
Can we split this method into two ones where the first one will make some preparation without read action and the second one will create expansion service after it in read action from data prepared by the first method?